src/kaleidoscope/device/dygma/Raise.cpp

```diff
@@ -0,0 +1,599 @@
+
+    // left under glow - 30
+    34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
+
+    // right underglow - 32
+    34 + LPH, 35 + LPH, 36 + LPH, 37 + LPH, 38 + LPH, 39 + LPH, 40 + LPH, 41 + LPH, 42 + LPH, 43 + LPH, 44 + LPH, 45 + LPH, 46 + LPH, 47 + LPH, 48 + LPH, 49 + LPH, 50 + LPH, 51 + LPH,
+    52 + LPH, 53 + LPH, 54 + LPH, 55 + LPH, 56 + LPH, 57 + LPH, 58 + LPH, 59 + LPH, 60 + LPH, 61 + LPH, 62 + LPH, 63 + LPH, 64 + LPH, 65 + LPH, XX, // last one is the Huble, never send to SLED
+  },
+  // ANSI
+  {
+    // left side - 32 keys includes LP: key 19 is missing for ANSI layout
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 20, XX, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 68, 69,
+
+    // right side - 36 keys includes LP
+    0 + LPH, 1 + LPH, 2 + LPH, 3 + LPH, 4 + LPH, 5 + LPH, 6 + LPH, 15 + LPH, 8 + LPH, 9 + LPH, 10 + LPH, 11 + LPH, 12 + LPH, 13 + LPH, 14 + LPH, 7 + LPH, 16 + LPH, 17 + LPH, 18 + LPH, 19 + LPH,
+    20 + LPH, 21 + LPH, 22 + LPH, 23 + LPH, 24 + LPH, 25 + LPH, 26 + LPH, 27 + LPH, 28 + LPH, 29 + LPH, 30 + LPH, 31 + LPH, 32 + LPH, 33 + LPH, 68 + LPH, 69 + LPH,
+
+    // left under glow - 30
+    34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
+
+    // right underglow - 32
+    34 + LPH, 35 + LPH, 36 + LPH, 37 + LPH, 38 + LPH, 39 + LPH, 40 + LPH, 41 + LPH, 42 + LPH, 43 + LPH, 44 + LPH, 45 + LPH, 46 + LPH, 47 + LPH, 48 + LPH, 49 + LPH, 50 + LPH, 51 + LPH,
+    52 + LPH, 53 + LPH, 54 + LPH, 55 + LPH, 56 + LPH, 57 + LPH, 58 + LPH, 59 + LPH, 60 + LPH, 61 + LPH, 62 + LPH, 63 + LPH, 64 + LPH, 65 + LPH, XX, // last one is the Huble, never send to SLED
+  }
+};
+
+/********* TWI *********/
+void twi_disable(void) {
+  Wire.end();
+}
+void twi_init(uint16_t clock_khz) {
+  Wire.begin();
+  Wire.setClock(clock_khz * 1000);
+}
+
+namespace kaleidoscope {
+namespace device {
+namespace dygma {
+
+/********* RaiseHands *********/
+
+struct RaiseHands {
+  static raise::Hand leftHand;
+  static raise::Hand rightHand;
+
+  static void setup();
+  static void initializeSides();
+
+  static uint8_t ansi_iso;
+
+  static void setSidePower(bool power);
+  static bool getSidePower() {
+    return side_power_;
+  }
+
+  static void keyscanInterval(uint16_t interval);
+  static uint16_t keyscanInterval() {
+    return keyscan_interval_;
+  }
+
+ private:
+  static uint16_t keyscan_interval_;
+  static bool side_power_;
+  static uint16_t settings_base_;
+};
+
+raise::Hand RaiseHands::leftHand(0);
+raise::Hand RaiseHands::rightHand(1);
+uint8_t RaiseHands::ansi_iso;
+bool RaiseHands::side_power_;
+uint16_t RaiseHands::settings_base_;
+uint16_t RaiseHands::keyscan_interval_ = 50;
+
+void RaiseHands::setSidePower(bool power) {
+  digitalWrite(SIDE_POWER, power ? HIGH : LOW);
```

Comment:
Hrm. As of now, we don't use digitalRead or digitalWrite -anywhere- outside of the examples (and the wip imago port.)  Historically, it's very, very slow due to a ton of extra bulletproof coding that's not typically desirable.

I don't know that we need to not do it. but it feels a little off

Reply:
I do not think we need the bullet-proofing here.

Reply:
I've looked into it, and... lets stick with `digitalWrite()`. It might be much slower, but we never use it on performance-critical codepaths. It is used in `setSidePower()`, which is called _very_ rarely. It is used in the side flashing code, which is the biggest user of `setSidePower()`. We also use it in `setup()`. None of these are performance critical.

And unlike on AVR, replacing `digitalWrite` is not all that easy on SAMD. See [here](https://forum.arduino.cc/index.php?topic=334073.msg2306350#msg2306350) and [here](https://forum.arduino.cc/index.php?topic=531268.msg3621078#msg3621078) for two scary examples.

I do not think that making our code so ugly is worth the performance gain here. If `digitalWrite` was used anywhere near a fast path, it would. But they aren't.